crates/matrix-sdk/src/encryption/mod.rs

```diff
@@ -876,46 +878,40 @@ impl Encryption {
     ///
     /// May reload the `OlmMachine`, after obtaining the lock but not on the
     /// first time.
-    pub async fn spin_lock_store(&self, max_backoff: Option<u32>) -> Result<(), Error> {
+    pub async fn spin_lock_store(
+        &self,
+        max_backoff: Option<u32>,
+    ) -> Result<Option<CryptoStoreLockGuard>, Error> {
         if let Some(lock) = self.client.inner.cross_process_crypto_store_lock.get() {
-            if lock.try_lock_once().await?.not() {
-                // We didn't get the lock on the first attempt, so that means that another
-                // process is using it. Wait for it to release it.
-                lock.spin_lock(max_backoff).await?;
-
-                // As we didn't get the lock on the first attempt, force-reload all the crypto
-                // state caches at once, by recreating the OlmMachine from scratch.
-                if let Err(err) = self.client.base_client().regenerate_olm().await {
-                    // First, give back the cross-process lock.
-                    lock.unlock().await?;
-                    // Then return the error to the caller.
-                    return Err(err.into());
-                };
+            match lock.try_lock_once().await? {
+                Some(guard) => Ok(Some(guard)),
+                None => {
+                    // We didn't get the lock on the first attempt, so that means that another
```

Comment:
Are we certain that we don't need to reload when we did acquire the lock on the first attempt? Isn't there a race where we might get the lock just as it was released by the other process?

Reply:
Indeed you're right! This is going to be more generally fixed by #2155 :-)